bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -1953,43 +1955,33 @@ void ReplicaImp::onMessage<ReplicaStatusMsg>(ReplicaStatusMsg *msg) {
   /////////////////////////////////////////////////////////////////////////
 
   else if ((msgViewNum == curView) && (!msg->currentViewIsActive())) {
-    auto sendViewChangeMsg = [&msg, &msgSenderId, this]() {
-      if (curView > 0 &&  // we only have ViewChangeMsg for View > 0
-          msg->hasListOfMissingViewChangeMsgForViewChange() &&
-          msg->isMissingViewChangeMsgForViewChange(config_.getreplicaId())) {
-        ViewChangeMsg *myVC = viewsManager->getMyLatestViewChangeMsg();
-        ConcordAssertNE(myVC, nullptr);
-        sendAndIncrementMetric(myVC, msgSenderId, metric_sent_viewchange_msg_due_to_status_);
+    auto sendViewChangeMsgs = [&msg, &msgSenderId, this]() {  // Send all View Change messages we have
+      if (curView > 0 &&                                      // we only have ViewChangeMsg for View > 0
+          msg->hasListOfMissingViewChangeMsgForViewChange()) {
+        for (auto *vcMsg : viewsManager->getViewChangeMsgsForView(curView)) {
+          if (msg->isMissingViewChangeMsgForViewChange(vcMsg->idOfGeneratedReplica())) {
+            sendAndIncrementMetric(vcMsg, msgSenderId, metric_sent_viewchange_msg_due_to_status_);
+          }
+        }
       }
     };
 
     if (isCurrentPrimary() || (repsInfo->primaryOfView(curView) == msgSenderId))  // if the primary is involved
     {
-      if (!isCurrentPrimary())  // I am not the primary of curView
-      {
-        sendViewChangeMsg();
-      } else  // I am the primary of curView
```

Comment:
Is this case captured by the final `else` case?

Reply:
The final else is for the case when the Primary is not involved. For the case where the Primary is involved we need to first send the New View message, if we are the current Primary and we have such, and then the View Change messages we have. Otherwise only send all View Change messages we have to the Primary.